### ISSUE TITLE
Problem with simple pressure controls

### ISSUE THREAD

**Issue by LRossman** [OpenWaterAnalytics/EPANET#642]
A simple pressure control that changes a pump's status to OPEN will fail to work if the input data set the pump status to CLOSED and did not supply a speed setting (so it defaults to 1). See the attached modified version of Net1.inp.
[net1-pressure_control.inp.txt](https://github.com/OpenWaterAnalytics/EPANET/files/6662430/net1-pressure_control.inp.txt)
In this example the pump has no speed setting specified but its initial status is set to CLOSED. The single simple control should turn the pump on if the pressure at node 32 is below 110. The EPANET solution shows node 32 at a pressure of  108.77 but the pump is still off.

**Comment by sergiosilvarubio** [comment]
Hello,
I think a condition is [missing](https://github.com/OpenWaterAnalytics/EPANET/blob/45c62f665677eacba5e57952cbd66215f303239f/src/hydraul.c#L531) (inside the for loop) that should verify that the node has a lower pressure than indicated, something like:

```c
double pressure = (hyd->NodeHead[n] - node->El) * pr->Ucf[PRESSURE];
if (control->Type == LOWLEVEL && pressure < control->Grade) reset = 1;
if (control->Type == HILEVEL && pressure > control->Grade) reset = 1;
```

but I have the doubt if it is possible to obtain the pressure value without having completed a hydraulic simulation.

PS: I have been using this tool for a short time but I am interested in starting to contribute to this great project, so I apologize if I have made a basic mistake.

**Comment by LRossman** [comment]
The `LOWLEVEL/HILEVEL` controls can be applied to both tanks and junctions. When applied to tanks the control is based on the water level in the tank which is always known before a solution for the next time period is found (i.e., it's a boundary condition). Therefore level controls for tanks can be evaluated and pipe/pump/valve status or setting changes can be made prior to finding the next hydraulic solution given the current levels of the tanks. That is what happens when the `controls()` function is called from `runhyd()`.

A level control based on a junction pressure needs to know what the pressure is before it can be evaluated. Therefore it cannot be included in the `controls()` function which is called prior to finding the next hydraulic solution using `hydsolve()`. Instead, within `hydsolve()` (in _hydsolver.c_), after a solution is found the pressure controls are checked by calling `pswitch()`. If the status/setting of any link changes, then the solution is updated to reflect the change.

**Comment by allenmlowe** [comment]
I tried this solution to solve this issue. It seems to work but needs to be checked further. I modified inithyd function in hydraul.c.

    // Initialize links
    for (i = 1; i <= net->Nlinks; i++)
    {
        link = &net->Link[i];
        link->ResultIndex = i;

        // Initialize status and setting
        hyd->LinkStatus[i] = link->Status;
        if (link->Status <= CLOSED) 
          hyd->LinkSetting[i] = 0.0;
        else
          hyd->LinkSetting[i] = link->Kc;

**Comment by Mariosmsk** [comment]
Yes, looks good. 

before
![image](https://user-images.githubusercontent.com/2945956/123509636-41ea5900-d67f-11eb-8a0c-6bf215461e9b.png)

after
![image](https://user-images.githubusercontent.com/2945956/123509600-136c7e00-d67f-11eb-956a-1a40db5e7d24.png)

**Comment by LRossman** [comment]
The proposed fix will cause problems when the link in the for-loop is not a Pump, since the Settings for Pipes and Valves should not be changed when their Status is set to `CLOSED`.

Since the bug is due to the way the input file is processed (and not to an API function), I think that a single line added to `changestatus()` in _input3.c_ (which processes the `[STATUS]` section of the input file) will fix it:
```
void changestatus(Network *net, int j, StatusType status, double y)
{
    ...
    else if (link->Type == PUMP)
    {
        if (status == ACTIVE)
        {
            link->Kc = y;
            status = OPEN;
            if (y == 0.0) status = CLOSED;
        }
        else if (status == OPEN) link->Kc = 1.0;
        else if (status == CLOSED) link->Kc = 0.0;     // new line
        link->Status = status;
    }
```

**Comment by allenmlowe** [comment]
I tried your solution @LRossman using the same sample. Pump 9 remained closed. How about only changing the link setting if it is a pump? 

// Initialize links
for (i = 1; i <= net->Nlinks; i++)
{
    link = &net->Link[i];
    link->ResultIndex = i;

    // Initialize status and setting
    hyd->LinkStatus[i] = link->Status;
    if (link->Status <= CLOSED || link->Type == PUMP) // new line
      hyd->LinkSetting[i] = 0.0;
    else
      hyd->LinkSetting[i] = link->Kc;

**Comment by LRossman** [comment]
My solution worked for me (applied to the latest EPANET-dev branch):
```
              Trial 18: relative flow change = 0.000019
                        maximum  flow change = 0.0123 for Link 21
                        maximum  head error  = 0.0000 for Link 31

              Pump 9 switched from closed to open
              Trial 19: relative flow change = 1.018615
              Trial 20: relative flow change = 1.036022
              Trial 21: relative flow change = 1.042666
              Trial 22: relative flow change = 0.941222
              Trial 23: relative flow change = 0.565100
              Trial 24: relative flow change = 0.160894
              Trial 25: relative flow change = 0.012089
              Trial 26: relative flow change = 0.000083
                        maximum  flow change = 0.0998 for Link 11
                        maximum  head error  = 0.0000 for Link 113

     0:00:00: Balanced after 26 trials
     0:00:00: Reservoir 9 is emptying
     0:00:00: Tank 2 is filling at 120.00 ft
     0:00:00: Pump 9 open
```

**Comment by allenmlowe** [comment]
You are right @LRossman, the latest Epanet-dev branch works with your latest modification. I probably made some modifications that made it fail with your latest modification. I will check mine further. Thanks.

**Comment by allenmlowe** [comment]
I modified the example in this issue and introduced a speed setting of 0.8. Should pump 9 still remain with a speed of 1.0 when it opens or should it use the speed setting, 0.8 when it opens? 

![Net1_Speed](https://user-images.githubusercontent.com/39788972/123572973-a7fdea00-d7ff-11eb-9df2-82a7fc55c6c9.jpg)
[net1-pressure_control.inp.txt](https://github.com/OpenWaterAnalytics/EPANET/files/6723013/net1-pressure_control.inp.txt)

**Comment by LRossman** [comment]
@allenmlowe the rules for changing STATUS/SETTING values for pumps are:

- if the STATUS becomes OPEN then the SETTING becomes 1.0
- if the STATUS becomes CLOSED then the SETTING becomes 0.0
- if the SETTING becomes > 0 then the STATUS becomes OPEN
- if the SETTING becomes 0 then the STATUS becomes CLOSED

These rules are enforced by `changestatus() `in _input3.c_ and by `setlinkstatus()` and `setlinksetting()` in _hydraul.c_. The latter two functions are called whenever a control implements an action or when the API function `EN_setlinkvalue()` is called.
